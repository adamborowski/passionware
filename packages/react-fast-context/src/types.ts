import { Draft } from 'immer';

export type Subscriber<T> = (state: T) => void;

export interface Store<T> {
  subscribe: (subscriber: Subscriber<T>) => () => void;
  update: (updater: (draftState: Draft<T>) => void) => void;
  replace: (state: T) => void;
  getState: () => T;
}
