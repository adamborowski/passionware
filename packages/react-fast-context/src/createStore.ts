import { Store, Subscriber } from './types';

export const createStore: <T>(initialState: T) => Store<T> = <T>(initialState: T) => {
  let state = initialState;
  const subscribers = new Set<Subscriber<T>>();

  const subscribe = (subscriber: Subscriber<T>) => {
    subscribers.add(subscriber);

    return () => {
      subscribers.delete(subscriber);
    };
  };

  const replace = (newState: T) => {
    state = newState;
    subscribers.forEach(subscriber => subscriber(state));
  };

  const getState = () => state;

  return { subscribe, getState, replace };
};
