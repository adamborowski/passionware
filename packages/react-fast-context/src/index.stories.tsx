import { ExampleComponent } from './index'
import React, { FC } from 'react'

export default { title: 'react-fast-context', component: ExampleComponent }

export const Default: FC<{ text: string }> = ({ text }) => (
  <ExampleComponent text={text} />
)
