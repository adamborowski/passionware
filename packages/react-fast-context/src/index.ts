export * from './createStore';
export * from './useStore';
export * from './StoreContext';
export * from './types';
export * from './useSelector';
export * from './useStoreContext';
