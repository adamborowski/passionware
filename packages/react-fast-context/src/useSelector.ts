import { useLayoutEffect, useReducer, useRef } from 'react';
import { useStoreContext } from './useStoreContext';

export type Comparator<T> = (a: T, b: T) => boolean;
const strictEqual: Comparator<unknown> = (a, b) => a === b;

export type Selector<State, SelectedState> = (state: State) => SelectedState;

// todo document
export const createUseSelector = <T, S>(selector: (state: S) => T, comparator?: Comparator<T>) => () => useSelector(selector, comparator);

export const useSelector = <State, SelectedState>(
  selector: Selector<State, SelectedState>,
  comparator: Comparator<SelectedState> = strictEqual
): SelectedState => {
  const [, forceRender] = useReducer((s: number) => s + 1, 0);
  const store = useStoreContext<State>();

  const latestValueRef = useRef<SelectedState>();
  const latestSelectorRef = useRef<Selector<State, SelectedState>>();
  const latestComparatorRef = useRef<Comparator<SelectedState>>();

  latestComparatorRef.current = comparator;

  if (selector !== latestSelectorRef.current) {
    latestValueRef.current = selector(store.getState());
    latestSelectorRef.current = selector;
  }

  useLayoutEffect(() => {
    const unsubscribe = store.subscribe(originalState => {
      const selectedState = latestSelectorRef.current?.(originalState);
      const latestValue = latestValueRef.current;
      if (
        (latestValue === undefined) !== (selectedState === undefined) ||
        (latestValue !== undefined && selectedState !== undefined && !latestComparatorRef.current!(latestValue, selectedState))
      ) {
        latestValueRef.current = selectedState;
        forceRender();
      }
    });

    return () => unsubscribe();
  }, [comparator, forceRender, store]);

  return latestValueRef.current!;
};
