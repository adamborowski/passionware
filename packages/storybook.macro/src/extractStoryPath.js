const path = require('path');
const _ = require('lodash');

module.exports = {
  extractStoryPath: (rootPackagePath, actualPackagePath, storyPath, packageName) => {
    const packageNameUnscoped = packageName.replace(/^@[A-z]+\//, '');

    const rootDirectory = path.normalize(path.resolve(rootPackagePath, '..'));
    const packageDirectory = path.normalize(path.resolve(actualPackagePath, '..'));

    const pathFromRootToPackage = path.normalize(path.relative(rootDirectory, packageDirectory));
    const pathFromPackageToFile = path.normalize(path.relative(packageDirectory, storyPath));

    const packagePath = pathFromRootToPackage.split(path.sep).slice(0, -1);
    const [storyKind, ...restOfPackagePath] = packagePath;

    const storyTitle = extractStoryTitleFromPathRelativeToPackage(pathFromPackageToFile, packageNameUnscoped);

    return `${_.capitalize(storyKind)}/${[restOfPackagePath.join(' '), packageNameUnscoped, ...storyTitle]
      .filter(segment => segment !== '')
      .join('/')}`;
  },
};

/**
 * For index stories it returns title to match parent folder story
 * @param path if ends at index - it will be shortened
 * @param packageNameUnscoped if path equals to packageNameUnscoped it also will be shortened
 * @return {*[]}
 */
const stripIndexStories = (path, packageNameUnscoped) => {
  const pathCopy = [...path];
  const length = pathCopy.length;
  const stripForIndexStories = pathCopy[length - 1] === 'index';
  const stripForStoriesWithNameSameAsPackageName = pathCopy.length === 1 && pathCopy[0] === packageNameUnscoped;
  const stripForStoriesWithNameSameAsFolderName = pathCopy[length - 1] === pathCopy[length - 2];
  if (stripForIndexStories || stripForStoriesWithNameSameAsPackageName | stripForStoriesWithNameSameAsFolderName) {
    pathCopy.pop();
  }
  return pathCopy;
};

/**
 * It gets story path relative to package root (ie. src/components/Foo/index.stories) and returns a user-facing path (ie. components/Foo)
 */
const extractStoryTitleFromPathRelativeToPackage = (pathFromPackageToFile, packageNameUnscoped) =>
  stripIndexStories(
    pathFromPackageToFile
      .replace(/\.stories(\.dev)?\.\w+$/, '')
      .split(path.sep)
      .slice(1),
    packageNameUnscoped
  );
