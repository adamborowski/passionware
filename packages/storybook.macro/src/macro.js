const { createMacro } = require('babel-plugin-macros');
const findPackage = require('find-package-json');
const { extractStoryPath } = require('./extractStoryPath');

module.exports = createMacro(storybookTitleMacros);

function storybookTitleMacros({ references, state, babel }) {
  const result = generateTitle({ state, babel });
  (references.title || []).forEach(referencePath => {
    referencePath.replaceWith(result);
  });
}

function generateTitle({ state, babel }) {
  const fileName = state.file.opts.filename;
  const packageFinder = findPackage(fileName);
  const packageJson = packageFinder.next();
  const rootJson = packageFinder.next();

  return babel.types.stringLiteral(extractStoryPath(rootJson.filename, packageJson.filename, fileName, packageJson.value.name));
}
