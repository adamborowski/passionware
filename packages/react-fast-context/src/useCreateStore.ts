import { useRef } from 'react';
import { createStore } from './createStore';
import { Store } from './types';

// todo add unit test to check if useCreateStore re-rendered updates the store
export const useCreateStore = <T>(state: T) => {
  const store = useRef<Store<T>>();
  if (!store.current) {
    store.current = createStore(state);
  } else {
    if (store.current.getState() !== state) {
      store.current.update(() => state);
    }
  }
  return store.current;
};
