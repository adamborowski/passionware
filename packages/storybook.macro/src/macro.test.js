import pluginTester from 'babel-plugin-tester';
import plugin from 'babel-plugin-macros';
import prettier from 'prettier';
import path from 'path';

jest.mock('./extractStoryPath', () => ({
  extractStoryPath: jest.fn().mockReturnValue('[generated title]'),
}));

pluginTester({
  plugin,
  snapshot: true,
  babelOptions: {
    filename: path.join(__dirname, '/deep/AwesomeComponent.stories.js'),
  },
  formatResult: result =>
    prettier.format(result, { trailingComma: 'es5', parser: 'babel' }),
  tests: {
    'no usage 1': 'import {title} from "../../src/macro";',
    'no usage 2': 'import anythingFooBar from "../../src/macro";',
    'basic usage - title as shortened object key-value form ': `
      import {title} from "../../src/macro";
      export default {title};
    `,
    'basic usage - title as object value ': `
      import {title} from "../../src/macro";
      export default {title: title};
    `,
    'basic usage - read as variable': `
      import {title} from "../../src/macro";
      const actualTitle = title;
    `,
    'advanced usage - story exports more data ': `
      import React from 'react';
      import {title} from "../../src/macro";
      export const Story1 = ()=> "A story";
      export default {title: title, decorators: [1,2,3]};
    `,
  },
});
