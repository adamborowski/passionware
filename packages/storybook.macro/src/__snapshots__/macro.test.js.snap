// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`macros advanced usage - story exports more data : advanced usage - story exports more data  1`] = `

import React from 'react';
import {title} from "../../src/macro";
export const Story1 = ()=> "A story";
export default {title: title, decorators: [1,2,3]};

      ↓ ↓ ↓ ↓ ↓ ↓

import React from "react";
export const Story1 = () => "A story";
export default {
  title: "[generated title]",
  decorators: [1, 2, 3],
};


`;

exports[`macros basic usage - read as variable: basic usage - read as variable 1`] = `

import {title} from "../../src/macro";
const actualTitle = title;

      ↓ ↓ ↓ ↓ ↓ ↓

const actualTitle = "[generated title]";


`;

exports[`macros basic usage - title as object value : basic usage - title as object value  1`] = `

import {title} from "../../src/macro";
export default {title: title};

      ↓ ↓ ↓ ↓ ↓ ↓

export default {
  title: "[generated title]",
};


`;

exports[`macros basic usage - title as shortened object key-value form : basic usage - title as shortened object key-value form  1`] = `

import {title} from "../../src/macro";
export default {title};

      ↓ ↓ ↓ ↓ ↓ ↓

export default {
  title: "[generated title]",
};


`;

exports[`macros no usage 1: no usage 1 1`] = `

import {title} from "../../src/macro";

      ↓ ↓ ↓ ↓ ↓ ↓



`;

exports[`macros no usage 2: no usage 2 1`] = `

import anythingFooBar from "../../src/macro";

      ↓ ↓ ↓ ↓ ↓ ↓



`;
